/*
 * Copyright 2021 Google LLC.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto2";

package temporian.proto;

// A processor defines a set of processing operations and holds all Event,
// Feature and Sampling instances.
message Processor {
  repeated Operator operators = 1;

  repeated Event events = 2;

  repeated Feature features = 3;

  repeated Sampling samplings = 4;
}

// Instantiation of an operator in a processor.
message Operator {
  // Unique identifier of the operator instance.
  optional string id = 1;

  // Key of the operator definition.
  optional string operator_def_key = 2;

  // Input event ids by key.
  repeated EventArgument inputs = 3;

  // Output event ids by key.
  repeated EventArgument outputs = 4;

  // Constructor attributes.
  repeated Attribute attributes = 5;

  message EventArgument {
    // Name of the argument.
    optional string key = 1;

    optional string event_id = 2;
  }

  message Attribute {
    // Name of the attribute.
    optional string key = 1;

    oneof type {
      int64 integer_64 = 2;
      string str = 3;
      string sampling_id = 4;
    }
  }
}

// Schema of an event
message Event {
  // Identifier of the event. Should be unique across all the events of the
  // processor.
  optional string id = 1;

  // Identifier of the sampling of the event.
  optional string sampling_id = 2;

  // Features in the event.
  repeated string feature_ids = 3;

  // Index of the event. Empty for global events.
  repeated string index = 4;
}

// Schema of a feature
message Feature {
  // Identifier of the feature. Should be unique in all the features in the
  // processor.
  optional string id = 1;

  optional Type type = 2;

  // Identifier of the sampling of the feature.
  optional string sampling_id = 3;

  // Index of the feature.
  repeated string index = 4;

  // Index of the event id who's operator creator also created this feature.
  optional string creator_event_id = 5;

  enum Type {
    UNDEFINED = 0;
    FLOAT = 1;
    INT64 = 2;
  }
}

// Schema of a sampling.
message Sampling {
  // Identifier of the sampling. Should be unique across all the samplings of
  // the processor.
  optional string id = 1;

  // Index of the sampling.
  repeated string index = 2;

  // Index of the event id who's operator creator also created this feature.
  optional string creator_event_id = 3;
}

// Definition of an operator. No two operators definition can have the same key.
message OperatorDef {
  // Unique identifier of the operator.
  optional string key = 1;

  // Inputs of the operator.
  repeated Input inputs = 2;

  // Outputs of the operator.
  repeated Output outputs = 3;

  // Arguments of the operator.
  repeated Attribute attributes = 4;

  // If true, the operator is a placeholder. That is, the user has to provide
  // the features of this operator output.
  optional bool place_holder = 5 [default = false];

  message Input {
    // String identifier of the input. Should be unique.
    optional string key = 1;

    // If true, the input is optional.
    optional bool is_optional = 3;
  }

  message Output {
    // String identifier of the output. Should be unique.
    optional string key = 1;
  }

  message Attribute {
    // String identifier of the atrribute. Should be unique.
    optional string key = 1;

    // What type of data is expected.
    optional Type type = 2;

    // If true, the attribute is optional.
    optional bool is_optional = 3;

    enum Type {
      UNDEFINED = 0;
      INTEGER_64 = 1;
      STRING = 2;
    }
  }
}
